lumicks/pylake/confocal.py

```diff
@@ -136,16 +137,29 @@ def _ordered_axes(self):
     def _to_spatial(self, data):
         raise NotImplementedError
 
-    def _image(self, color):
-        if color not in self._cache:
-            photon_counts = getattr(self, f"{color}_photon_count").data
-            self._cache[color] = reconstruct_image_sum(photon_counts, self.infowave.data, self._shape)
-            self._cache[color] = self._to_spatial(self._cache[color])
-        return self._cache[color]
-
-    def _timestamps(self, sample_timestamps):
-        timestamps = reconstruct_image(sample_timestamps, self.infowave.data, self._shape, reduce=np.mean)
-        return self._to_spatial(timestamps)
+    @cachetools.cachedmethod(operator.attrgetter('_cache'))
+    def _from_infowave(self, channel):
+        if not channel in ("timestamps", "red", "green", "blue"):
+            raise ValueError(f"{channel} channel is not recognized.")
+
+        kwargs = {}
+
+        if channel == "timestamps":
+            # Uses the timestamps from the first non-zero-sized photon channel
+            for channel_data in (getattr(self, f"{color}_photon_count").timestamps 
+                                 for color in ("red", "green", "blue")):
+                if len(channel_data) != 0: 
+                    break
+            else:
+                raise RuntimeError("Can't get pixel timestamps if there are no pixels")
+            fcn = reconstruct_image
+            kwargs["reduce"] = np.mean
+        else:
+            channel_data = getattr(self, f"{channel}_photon_count").data
+            fcn = reconstruct_image_sum
```

Comment:
Given how many differences there are between these two paths, I feel like it might actually be simpler to split it into 2 separate functions (colors and timestamps) rather than force them together to reuse just 1 or 2 lines.

I guess that you joined the 2 functions in order to share the single cache but I don't think that this is strictly necessary.

Reply:
True. I was trying to simplify the internals (everything that uses the cache calls the same method) but the method itself did indeed get a bit convoluted. Turns out you can use the same cache for different methods though if they have the same signature